src/core/mesh/qgsmeshlayer.h

```diff
@@ -0,0 +1,224 @@
+ *      "2.0, 2.0 \n" \
+ *      "3.0, 2.0 \n" \
+ *      "2.0, 3.0 \n" \
+ *      "1.0, 3.0 \n" \
+ *      "---"
+ *      "0, 1, 3, 4 \n" \
+ *      "1, 2, 3 \n"
+ *    );
+ *    QgsMeshLayer *scratchLayer = new QgsMeshLayer(uri, "My Scratch layer", "memory_mesh");
+ * \endcode
+ *
+ * \subsection mdal MDAL data provider (mdal)
+ *
+ * Accesses data using the MDAL drivers (https://github.com/lutraconsulting/MDAL). The url
+ * is the MDAL connection string. QGIS must be built with MDAL support to allow this provider.
+
+ * \code
+ *     QString uri = "test/land.2dm";
+ *     QgsMeshLayer *scratchLayer = new QgsMeshLayer(uri, "My Scratch Layer",  "mdal");
+ * \endcode
+ *
+ * \since QGIS 3.2
+ */
+class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
+{
+    Q_OBJECT
+  public:
+
+    /**
+     * Constructor - creates a mesh layer
+     *
+     * The QgsMeshLayer is constructed by instantiating a data provider.  The provider
+     * interprets the supplied path (url) of the data source to connect to and access the
+     * data.
+     *
+     * \param path  The path or url of the parameter.  Typically this encodes
+     *               parameters used by the data provider as url query items.
+     * \param baseName The name used to represent the layer in the legend
+     * \param providerLib  The name of the data provider, e.g., "mesh_memory", "mdal"
+     */
+    explicit QgsMeshLayer( const QString &path = QString(), const QString &baseName = QString(), const QString &providerLib = "mesh_memory" );
+    //! Dtor
+    ~QgsMeshLayer() override;
+
+    //! QgsMeshLayer cannot be copied.
+    QgsMeshLayer( const QgsMeshLayer &rhs ) = delete;
+    //! QgsMeshLayer cannot be copied.
+    QgsMeshLayer &operator=( QgsMeshLayer const &rhs ) = delete;
+
+    //! Return data provider
+    QgsMeshDataProvider *dataProvider() override;
+
+    //! Return const data provider
+    const QgsMeshDataProvider *dataProvider() const override SIP_SKIP;
+
+    /**
+     * Returns a new instance equivalent to this one. A new provider is
+     *  created for the same data source and renderers are cloned too.
+     * \returns a new layer instance
+     */
+    QgsMeshLayer *clone() const override SIP_FACTORY;
+
+    //! Returns the extent of the layer.
+    QgsRectangle extent() const override;
+
+    /**
+     * Return new instance of QgsMapLayerRenderer that will be used for rendering of given context
+     */
+    virtual QgsMapLayerRenderer *createMapRenderer( QgsRenderContext &rendererContext ) SIP_FACTORY;
+
+    //! Return the provider type for this layer
+    QString providerType() const;
+
+    //! return native mesh (nullprt before rendering)
+    QgsMesh *nativeMesh() SIP_SKIP {return mNativeMesh;}
+
+    //! return triangular mesh (nullprt before rendering)
+    QgsTriangularMesh *triangularMesh() SIP_SKIP {return mTriangularMesh;}
+
+    //! Returns a line symbol used for rendering native mesh.
+    QgsSymbol *nativeMeshSymbol() {return mNativeMeshSymbol;}
```

Comment:
I'm not sure about having symbol and toggleTriangularMeshRendering here - those seem to be more of mesh renderer properties, and I think belong in their own separate class.